### DISCUSSION TITLE
Explicit Filecoin Plus verified deal premium (was: subsidy)

### DISCUSSION THREAD

**Discussion by anorth** [filecoin-project/FIPs#243] [Enhancements - Technical]
This proposal is an idea of mine to reduce coupling between the storage power mechanism and deal market. Coupling between these two limits design freedom around problems like supporting exponential growth in storage capacity or making deals much more scalable.

**⭐️Update – January 2022**
This proposal is now primarily motivated by programmability of Filecoin, and storage deals in particular, once the FVM enables user-programmable contracts. See discussion in #241 for a bigger picture background. While greater scalability of storage accounting provided the initial spark, it's no longer the primary target, just a bonus. The [first draft](https://github.com/filecoin-project/FIPs/pull/263) of the FIP dropped a bunch of changes that were only about storage accounting.

I have left the body of this post intact to preserve context for the subsequent context, but [re-expressed the motivation here](https://github.com/filecoin-project/FIPs/discussions/243#discussioncomment-1959954).

---
The description below is deprecated. See the [PR for the first draft of this FIP](#263) for an explanation of this proposal.

## Background
Filecoin network storage growth will eventually be limited by the chain computation costs associated with maintaining storage if a linear amount of state is accessed or retained per sector (where sectors are a fixed size). The Neutron proposal (#119) resolves this for committed-capacity sectors, but sectors with deals still have per-sector (and per-deal) metadata. Per-sector on-chain data includes the sector->deal mapping and per-sector heterogeneous weight, pledge and penalty values. These are all required because **verified deals alter a sector’s power** as a means of adjusting reward distribution.

Maintaining the facility for heterogenous sectors constrains freedom for more scalable designs, even though most sectors don’t have deals. The premise that most sectors don’t have deals is not something to rely on, though. In the long term we aim for many more deals and a much greater proportion of committed storage to be in active use. 

## Goals
This proposal aims to reduce deal market limitations on the  scale and efficiency of onboarding and maintaining exponentially larger amounts of capacity.

- Remove per-sector on-chain information from state
- Normalise sectors as far as possible, enabling summarised/aggregate accounting with sub-linear state
- Decouple network security from reward distribution policy

These goals are to be sought within existing product and crypto-economic constraints around sound storage and economic security.

## Design ideas
The current storage power mechanism assigns a different power value to equal-size sectors based on the size and duration of Filecoin Plus (FIL+) verified deals. This is a means of incentivising useful storage, delegating the definition of “useful” to an off-chain notary. The incentive for storing verified deals is the increased block reward that may be earned from the increased power of those sectors, despite constant physical storage and infrastructure costs. In essence, **the network subsidises useful storage by taxing the other power**.

Storage power has two different roles, currently coupled. One is to secure the network through raising the economic cost of some party maintaining a significant fraction of the total power, and the other is to determine the distribution of rewards. The block reward is split between rewarding security and subsidising useful storage. This coupling theoretically means the verified deal power boost reduces the hardware cost to attack the network by a factor of 10, if colluding notaries would bless all of a malicious miner’s storage (this is an impractical attack today, though).

This **subsidy could be made much more direct**, reducing complexity and coupling between the storage market and individual sectors, and clearly separating security from reward distribution policy.

### Uniform sector power
Every sector has uniform power corresponding to its raw committed storage size, regardless of the presence of deals. This **removes the concepts of sector quality and quality-adjusted power**, and makes all bytes equal in the eyes of the power table. This would remove the DealWeight and VerifiedDealWeight fields from SectorOnChainInfo. Network power and committed storage are now the same concept and need not be tracked separately by the power actor.

Uniform power also means that all similarly-sized sectors committed at the same epoch would require the same initial pledge. Similarly the expected day reward and storage pledge (parameters to a possible future termination fee calculation) depend only on activation epoch. The complicated values maintained to track penalties for replaced CC sectors become unnecessary. Historical pledge/reward values may be maintained just once for the network by the power and reward actors. We currently store each of these numbers in chain state some ~500 times per epoch (@ 50PiB/day growth).

With uniform sector power, the power of groups of sectors may be trivially calculated by multiplication. Window PoSt deadline and partition metadata no longer need to maintain values for live, unproven, faulty and recovering power, but the sector number bitfields remain. Processing or recovering from faults does not require loading each sector’s specific values. The complexity of code and scope for error in these various derived data is much reduced.

This ability to aggregate by multiplying becomes instrumental to Neutron-like ideas around hierarchical storage accounting with deals. **Without this, supporting partial faults requires on-chain metadata about the individual sectors (with deals) that comprise a supersector**, restoring a linear-in-power network cost.

Note that this proposal does leave the per-sector deal IDs on chain. After this proposal and Neutron, this would be the only piece of per-sector data retained.

A miner’s chance of winning the election at each epoch is proportional to their share of the raw byte power committed to the network and non-faulty. Winning the consensus reward remains a lottery.

### Market actor tracks deal space-time
The market actor maintains:
- a current total of active verified deal space (optionally unverified, too);
- a short list of “reward periods”, each aggregating a period of, say, 24 hours, and comprising:
    - a table of verified deal space-time totals provided by each miner during the period;
    - a record of the total deal subsidy earned during the period (see below)

As payments are periodically processed for each deal (currently every 100 epochs), the market actor adds the deal’s size multiplied by the update period epochs to the provider’s space-time tally for the current reward period.

After a reward period completes, the ratio between verified deal space-time of each provider gives their share of the deal subsidy to be claimable. A miner can claim a corresponding share of the total deal subsidy earned at any point up until the reward period expires (e.g. 60 days). Upon claiming a deal subsidy, the reward funds are inserted into the miner’s vesting schedule, alongside block rewards.

### FIL+ subsidy paid to market actor
At every epoch, the reward actor queries the power actor for the total network power (= storage) and the market actor for the total verified deal space. The total block reward to be paid is then split into the power reward and the verified deal subsidy according to the ratio of storage to 9*deal space. The block reward is paid as usual, and the deal subsidy is sent to the market actor.

## Discussion
The primary motivation for this proposal is to remove per-sector account-keeping metadata in order to unlock exponential scaling of storage. It changes reward behaviour in a couple of ways that we must verify as being beneficial.

- Earning the verified deal subsidy doesn’t depend on winning blocks. This is a deviation from the current protocol that requires a miner to win a block (with a Winning PoSt) in order to claim rewards. This is **more reliable for smaller parties** that might win only occasional blocks, and might suffer from less robust blockchain node infrastructure and connectivity. 
- The market actor does not track temporary faults in sectors: a provider is eligible for the full client payment if the corresponding sector faults, so long as it is recovered soon thereafter. The power table does track temporary faults, suspending rewards until a sector is recovered (necessary for economic arguments about security). This change moves the deal subsidy from being subject to suspension during faults to the same more tolerant treatment of deals. Verified deals no longer subvert the economic security arguments, but it’s not obvious that this makes the more tolerant treatment acceptable.
If not, we might need to communicate faults to the market actor, making them a bit more expensive and exposing the market actor to a concept of sectors it’s currently mostly abstracted from.
- We probably need a separate pledge for the deal subsidy, so that collateral remains proportional to future earnings. This pledge would be about (verified) deals, rather than sectors, because this proposal separates those reward streams. This pledge would likely be owned by the market actor and forfeit when a deal defaults.
- A potentially tricky situation to analyze is what happens when the power reward is significantly less than the deal subsidy. Does it create any unwanted incentives for the block producing miner to deviate from the protocol (assuming they don’t have many deals)?

**Comment by anorth** [comment]
[For a follow-up proposal, I think the final step in decoupling storage power from deal market is to move the sector->deal mapping from sector metadata into some equivalent but reversed deal->sector mapping in the market actor]

**Comment by kaitlin-beegle** [comment]
Migrating this to the discussion channel until the FIP is ready for community stewardship!

**Comment by Fatman13** [comment]
Are there any efforts pushing this forward rn?

**Reply by anorth** [reply]
Yes. I am actively working on a draft FIP document, and prototyping implementation in the Go actors in order to tease out some of the details that are difficult to think through.

Thanks for the prod. I will endeavour to publish my draft soon, though it will still have a few decisions to be made.

**Comment by anorth** [comment]
I have published a first draft of this FIP: https://github.com/filecoin-project/FIPs/pull/263

**Comment by AxCortesCubero** [comment]
As we understand, the proposal is oriented towards enhancing the technical capacity for network growth. The proposed changes will simplify the calculations that need to be done at each epoch, enabling faster network growth. This implementation, however, has direct cryptoeconomic consequences that we would need to understand further. 

The first cryptoeconomic concern we have identified is that this proposal alters the existing incentive structure. The total reward given out to miners has to be re-arranged into a part that involves only the raw power, and a deal subsidy that involves the QAP. We have done [some work](https://hackmd.io/@R02mDHrYQ3C4PFmNaxF5bw/SJ61McoOt) on how these two pieces of the reward could be arranged, such that the total reward remains the same as it is now, however there is some freedom of different possible choices on how to redistribute this reward. There is also significant work to be done on understanding and designing the corresponding changes in the initial pledge and penalty structure.

Even if reward is rearranged such that the total expected miner reward is not changed, the consensus power will change, as it would now be determined only by the miner’s raw power, instead of their QAP. This is unclear if this is a desirable property, as the loss in consensus power, may disincentivize miners from seeking to store verified deals.

Lastly, it is not apparent to us that the added growth capacity is urgently needed at this moment. We believe that network growth is not currently limited by this technical capacity. Given the scale of the change involved in this FIP (collateral, reward, and penalty) and the lack of clarity in economic benefits (or impact unknown without more detailed analysis), we (CryptoEconLab) recommend de-prioritizing it till later.

**Reply by anorth** [reply]
> As we understand, the proposal is oriented towards enhancing the technical capacity for network growth. 

This is incorrect. I can understand how you reached this view though because the original spark for this idea was indeed scale. I've commented below on how this thinking has changed.

> alters the existing incentive structure

It alters the _mechanism_ but aims to leave the actual incentives intact. Yes, there is work to be done detailing, modelling, and gaining confidence in the mechanism and its behaviour, which is what this discussion and the FIP process are for.

> loss in consensus power, may disincentivize miners from seeking to store verified deals

This is a good point and I've now incorporated it into the the design rationale of the draft. On the assumption that the total rewards remain the same (even with more favourable timing), it's not clear for what other reasons a provider may wish to stack consensus power. Reputation? To manipulate the blockchain? Removing abuse of the FIL+ system as a vector for gaining consensus power (vs the economically-hard work of proving storage) is explicitly called out as a benefit, here.

> it is not apparent to us that the added growth capacity is urgently needed at this moment

This is no longer relevant. This change is about programability.

**Reply by Fatman13** [reply]
> This is unclear if this is a desirable property, as the loss in consensus power, may disincentivize miners from seeking to store verified deals.

Current QAP scheme along with fil+ program didn't necessarily do a great job in incentivizing the on-boarding of verified deals either,  given the growth trajectory of verified deals and the network. ([～9Pib](https://filplus.info/) compared with the 14EiB of the network). The network needs tools like fvm to break the current paradigm where programs like f+ tried and failed to on-boarding a significant amount of deals onto the network over the timespan of last year. And as the proposal said this change is critical to fvm's future programability. It is apparent that having this proposal before fvm would mitigate lots of technical challenge the team may have if this were to be implemented after fvm. To have a stabler actors before fvm in general should be desired.

**Reply by jennijuju** [reply]
> do a great job in incentivizing the on-boarding of verified deals either, given the growth trajectory of verified deals and the network. (～9Pib compared with the 14EiB of the network)

I think a more fair comparison to consider whether “verified deal incentives are good or not” should be the amount of verified deals vs the amount of the regular deals, instead of the total network power - the later  is more related to the q of “whether making deals is incentive or not comparing to on boarding CC sectors in general ” imho

**Reply by AxCortesCubero** [reply]
> It alters the _mechanism_ but aims to leave the actual incentives intact. Yes, there is work to be done detailing, modelling, and gaining confidence in the mechanism and its behaviour, which is what this discussion and the FIP process are for.

The way it alters the mechanism I think is more subtle, and thus not that clear what is the effect. I mean the most basic aim would be to leave the total amount of individual miner reward, unchanged, as it is now. Yet there might be a change in the timing of how these rewards are given. More importantly I think there are more specifications to be made, about how rewards would be given out, even if total reward is the same. Some of the choices may be even "better" than what we have, but requires more effort to study in detail what is the best choice, and how that best choice compares with the current structure.

> This is a good point and I've now incorporated it into the the design rationale of the draft. On the assumption that the total rewards remain the same (even with more favourable timing), it's not clear for what other reasons a provider may wish to stack consensus power. Reputation? To manipulate the blockchain? Removing abuse of the FIL+ system as a vector for gaining consensus power (vs the economically-hard work of proving storage) is explicitly called out as a benefit, here.

This is another point that is less straightforward to answer. For miner reward, it is clear that it is a reward, every rational miner would want to maximize their reward. Consensus power is less objective, it may be seen as a reward for some, or a burden to others (so we would also be making changes to the distribution of burden). In any case, however the individual miner sees it, they know how much consensus power to expect currently, and this would be a major change.

> This is no longer relevant. This change is about programability.

Point taken, I need to think more about this aspect.

**Comment by rjan90** [comment]
I think this proposal is great for the long term, and that it will be an improvement for storage providers to decouple FIL+ subsidy from raw-power consensus rewards, and that this will enable more providers (and a more diverse set) in the Filecoin storage market long term. But I also agree with @AxCortesCubero that this FIP seems alter the current incentive structure so more research should definitely be done in that space.

_Reasoning why I think this is good:_ Currently, storage providers income is heavily dependent on block rewards (probably in the >99% area). Since the FIL+ system (QAP) is coupled to block-rewards, it also hinders smaller providers who are limited by resources (either if that is money/space/electricity, or something else which cause their size to be smaller) to be able to get a steady and predictable reward/income since its coupled to block rewards. As an example, if you are a storage provider with only 50TiB raw-storage (and no FIL+ deals), you are currently looking at getting on average a block approximately every month, and with a 99.9%-probability, way longer then that. This skew will continue as long as the network grows and will hinder smaller participants to join. Decoupling FIL+ rewards from the block rewards might make it more reasonable for smaller providers in more remote places to join the network.

**_Hypothetical scenario:_** Someone wants to start a storage provider in a remote area like Svalbard, and are limited by power, available space or some other limiting factor to their size. I would say that even if this provider can´t become a 1+ PiB storage provider, they offer some very interesting properties with regards to its location to the Filecoin network. Decoupling FIL+ subsidy would possibly enable this provider a more predictable stream of income rather then relying on block rewards which might be out of this providers scope when the network becomes larger.

**Comment by anorth** [comment]
The original motivation for this proposal (six months ago) was scalability, but _that is not the motivation today_. I have left the original post intact to give retain context for the subsequent comments, but the motivation and part of the mechanism is out of date.

This proposal is motivated primarily by programmability, setting the stage for innovation in storage deal functionality and marketplaces once the FVM enables user-programmable contracts. Features like deal extension, multi-sector deals, re-negotiation, deal transfer, capacity deals, repair markets, etc should all be possible in smart contracts. But even if we had the FVM today, we wouldn't be able to realise most of them due to the coupling between (verified) deals and individual sector power, and between the miner actor and the built-in storage market. 

Breaking this coupling will also have other benefits. It will significantly speed up future development of both the storage power accounting and future deal markets, support much more scalable data representations, remove the storage market from consensus-critical risk surface, and move subsequent market development into the realm of smart-contract upgrades, rather than network-wide consensus protocol upgrades. This direction is thus motivated by speed, scale, flexibility and programmability.

See, and please contribute to, #241 for discussion of this broad motivation.

**Comment by jennijuju** [comment]
Understood that the subsidy will be reaccounted - but Seems like the proposal is to remove QAP for existing verified deals as well - is that so?

**Reply by anorth** [reply]
Yes that's right, the QAP will be converted to verified deal accounting.

**Reply by jennijuju** [reply]
I understand the technical needs here and I think the proposal makes a lot of sense on decoupling market and power/rewards actor. however, I think the sudden power change will impact enterprise SP’s biz as mostly they promote their service base on how much QAP they have served - would love to hear more from the fil+ community on how they feel about this.

**Reply by Fatman13** [reply]
> In my opinion, business failure is exactly the kind of risks that storage providers are paid to manage.

https://github.com/filecoin-project/FIPs/discussions/255#discussioncomment-1903260

**Reply by jennijuju** [reply]
@Fatman13 could you elaborate more on how your qoute is relevant to my concern, are you suggesting you consider that as a risk that SP should manage? 

I think a sudden power changes due to how the power is accounted from the protocol  is *very* different from service failure due to “force majority”  or bad operation.

**Reply by Fatman13** [reply]
> however, I think the sudden power change will impact enterprise SP’s biz as mostly they promote their service base on how much QAP they have served

Sorry, let me clarify.

I am generalizing the worst outcome of how QAP could impact SP's business and used the comment for the potential passing of this proposal shouldn't be depend on how fil+ community feel about if they could use QAP to promote their business.

**Reply by jennijuju** [reply]
i disagree - It should be a part of product consideration.

**Reply by jennijuju** [reply]
from https://github.com/filecoin-project/FIPs/pull/263#issuecomment-1011253388 
> change something that works and is finally accepted by the SP's.

> Also, as it is right now, every SP has at least a clear visible thing "Raw power vs Adjusted Power" and clear on-chain 10x growth.

@anorth is it possible to keep QAP as a storage power concept, while account the subsidy differently? For example, only use QAP for reward accountant instead of using it for winning election?

**Reply by anorth** [reply]
No, retaining the QAP concept for some reasons would remove all the decoupling/programmability benefits of the proposal.

It's easy to compute (off-chain) what the QAP would have been, if it still existed, but since the chain no longer cares about it I don't think that would stick.

**Reply by jennijuju** [reply]
That makes sense - thanks!

**Comment by jennijuju** [comment]
have we done any analysis on how this will impact the reward distribution(higher or lower probabilities on winning counts based on the current network power) for SPs that are  serving verified deals atm?

**Reply by anorth** [reply]
Not in detail. I'm working on that kind of analysis now, so thanks for the question. Note that for smaller SPs who aim to have mostly verified power (i.e. high QAP to raw power ratio), their probabilities of reward will improve from being a lottery to being ~certain and smooth.

**Reply by AxCortesCubero** [reply]
I think the more important point is that it is underspecified what the change to the reward distribution would be. There are choices still to be made to be made. The most minimal approach would be to keep the total miner reward unchanged by balancing the block reward and deal subsidy. There are however multiple ways in which this can be acheived, so it would first be needed to choose what would be the best strategy, and then analize the impact of this change vs the current structure. This is why I say it requires a more research efforts.

**Comment by zixuanzh** [comment]
Reward is just one part of the economic structure that this FIP is touching. However, when we change the reward, we also need to update the pledge and penalty calculation and understand how different parts interact with each other on the macro level. We can do our best to keep the economic structure, policy and incentives intact but this is also a significant undertaking. Given when we are for the network (there is a major economic structural shift since HyperDrive; the network is still far from an equilibrium/stationary state; and storage demand is just picking up (FIL+)), we cautioned against making large economic changes to the network.

I would also like to push back on the assumption that storage market innovation is blocked by the current design. People can do DataDAOs, storage bounties (by extension repair markets), deal renewals/extensions (just do a new one/snap it), user-defined deal markets with the current design and storage state bridges to other smart contract systems. What we lack as an ecosystem are more entrepreneurs and builders who understand the stack deeply to make it happen and successful examples of such applications (think CryptoKitties for NFTs) to validate business assumptions and drive adoption.

**Reply by Fatman13** [reply]
Wether entrepreneurs and builders are coming into the ecosystem is also somewhat depending on how well the cryptocurrency performs. For example, CryptoKitties was released at the height of eth rally during 2017.

**Comment by anorth** [comment]
Progress update: I have re-expressed much of the motivation for this proposal in the draft PR. We've been working on validating that the rewards remain stable etc.

I also rebranded the proposal to the Filecoin Plus premium, rather than subsidy, reflecting the positive incentive to valuable data that the system is aiming at, and how this proposed mechanism enhances that.

This proposal is just one of the changes necessary to unlock storage market innovation with the FVM. As we've been working through the other pieces, it seems like maybe there's more initial value to be unlocked in re-architecting the actor APIs to provide the hooks for alternative storage markets, even if they'd be restricted to write-once storage. I'll be drafting up an outline of that shortly, but wanted to flag here that it seems probably that we'd sequence the FIL+ premium *after* that re-architecture work. So I'll probably pause work on this proposal for a couple of months.
